### ISSUE TITLE
Parallization of gap_fit with openmp

### ISSUE THREAD

**Issue by yuxzhou** [libAtoms/QUIP#365]
Hi users and developers of QUIP,

I am using gap_fit of gfortran openmp version for my potential fitting and found that the CPU usage didn't reach the maximum provided, i.e., I required 48 cores (1 node) for the fitting in my local cluster and the max %CPU (seen by using _top_ in the shell) stopped at ~2000%. And at most of time, the %CPU is only around 100% (as shown below), which seems to only run in serial. However, I guess it should have reached ~4800%, for most of parts of the fitting, right? Do you think it a problem of code compilation or the cluster itself?
```
 PID    USER    PR  NI    VIRT    RES        SHR S    %CPU   %MEM     TIME+       COMMAND
 4222  *****     20   0   131.8g  130.4g   2996 R     99.7       8.6         2138:26        gap_fit

```
Best,

**Comment by gabor1** [comment]
if you look at the output of gap_fit, the timing information contains both wall time and cpu time. only the most expensive operations are parallelised, so you will indeed find that some things are serial. For small problems, they can even dominate the runtime, but for large databases and large descriptor sizes and basis sizes (nsparse), the parallelisation is pretty good.

**Comment by yuxzhou** [comment]
Thank you! In fact, I'm fitting a relatively large database using nsparse=5000, and I suppose the parallelization should be pretty good? Maybe I should check the infrastructure of cluster and re-compile the code again.

**Comment by gabor1** [comment]
What about actual runtime? Have you checked how long it takes compared to using fewer cores?

— Gábor

Gabor Csanyi
Professor of Molecular Modelling
Engineering Laboratory
University of Cambridge

‐‐‐‐‐‐‐ Original Message ‐‐‐‐‐‐‐

**Comment by noambernstein** [comment]
My experience with large numbers of cores (on Intel CPUs), say 56 or 72, is
that it’s actually slower to use all of them. I typically set
OMP_NUM_THREADS to at most half the physical cores, maybe even less.

**Comment by mcaroba** [comment]
One way to check what's happening during runtime is to ssh onto the compute node you're using and run htop. @noambernstein Our experience is that we can use all cores efficiently if we adjust the OpenMP scheduling. @albapa kindly made the openmp_chunk_size variable tunable from the command line. You can use very large values, for TB sized covariance matrices we can effectively set it to 10000 (the default of 1 is very inefficient).

**Comment by bernstei** [comment]
If openmp_chunk_size is so important, should we change the default from 1 to at least something moderate?

**Comment by mcaroba** [comment]
I think a default of at least 100 would make sense.

**Comment by Sideboard** [comment]
A small test with 11 Si structures, which takes about 11 min for serial, suggests a chunk size between 1000 and 10000.
![image](https://user-images.githubusercontent.com/1591390/144286413-a1df48f0-168c-417a-9fbd-c4e87347133c.png)

**Comment by mcaroba** [comment]
@Sideboard Is this the speedup for covariance matrix construction? Or the whole calculation? The total time is affected by the serial parts of the code and also there's the final linear algebra calls that can benefit from a threaded BLAS/LAPACK installation. Also, how many physical cores do you have in this machine? These are all details that would shed light on whether it makes sense to fully commit the physical cores (i.e., OMP_NUM_THREADS = num_physical_cores) or not. The question of whether openmp_chunk_size = 1 is a bad default is quite clear IMO, on the other hand.

**Comment by Sideboard** [comment]
These numbers are taken with second accuracy directly from gap_fit output (stop time minus start time), so it's the whole calculation including serial parts. They were run on a cluster of identical nodes with 48 physical cores each, so the right-hand side points are a full OMP commit. I was using the queue, so some random timing noise will be included as well; when looked at these timing last (for the default chunk_size=1 only) the speedup decreased after about 12 threads. This was not meant as a comprehensive test, otherwise we should also look at larger systems.

**Comment by mcaroba** [comment]
Thanks, I will try to make a comprehensive test with a bigger database next week. I will add one last bit. Besides setting OMP_NUM_THREADS we also set:

    export OMP_STACKSIZE=1G
    export OMP_DYNAMIC=false
    export OMP_PROC_BIND=true

I don't remember how/why we came to the conclusion that these were good choices. I might also try to check if changing these makes any difference (especially the last export).

**Comment by Sideboard** [comment]
@mcaroba Did you get around to run some larger tests?

**Comment by mcaroba** [comment]
@Sideboard No, sorry! I also went "offline" for a few days (and will continue at very low pace until the new year), but thanks for the reminder that I promised to do this. I am finishing some database computations and will be training production force fields in a couple of weeks, with a full node's worth of memory. I can do these tests at the same time then. But please do (angrily) ping me again later if you're still interested in the answer (my work is fueled by angry reminders, it's what keeps me going!).

**Comment by mcaroba** [comment]
@Sideboard I know it has been a long time, but I finally managed to get around to fit these big-database potentials (total memory usage of circa 3 TB). I tried 3 different chunk sizes and this is the time it took with 48 CPU cores (and the same number of OpenMP threads) to do the covariance matrix calculation for the SOAP (soap_turbo) part:

Chunk size; Time in hours
  1000    8.9
 10000    6.8
100000    6.8

So there is clearly an optimum and it is quite large for production calculations.

**Comment by Sideboard** [comment]
FYI: With https://github.com/libAtoms/QUIP/pull/484 we have introduced the special input `0` for the chunksize and made it the default. With this the loops get split into `ceil(n_items / n_threads)` chunks, so a single chunk per thread.

I think this issue can be closed.

**Comment by Sideboard** [comment]
Here is the result (0, dark blue) for small test:
![OMP_0](https://user-images.githubusercontent.com/1591390/204863190-8983837f-205c-47b4-9324-c786ccc12bbf.png)
